### PR TITLE
Make all integration tests safe to run in parallel

### PR DESCRIPTION
**Pull request by corylanou**
Packages test in parallel by default, and with the recent addition of integration tests in other packages, this is now creating intermittent port collision, and hence, test failures.

This change also required the following changes:

- Track Graphite listeners so we could close them properly and assign `0` port for testing
- Track OpenTSDB listeners so we could close them properly and assign `0` port for testing
- Change they way we do `default` config values for Graphite/OpenTSDB endpoints as to allow for testing.  All other `default` styles should also likely move to this new style.

## Before the change:

```
$ time go test ./... -p=1
ok      github.com/influxdb/influxdb    3.597s
ok      github.com/influxdb/influxdb/admin      0.011s
ok      github.com/influxdb/influxdb/client     0.009s
ok      github.com/influxdb/influxdb/cmd/influx 4.647s
ok      github.com/influxdb/influxdb/cmd/influxd        111.165s
ok      github.com/influxdb/influxdb/collectd   0.007s
ok      github.com/influxdb/influxdb/graphite   0.006s
ok      github.com/influxdb/influxdb/httpd      1.539s
ok      github.com/influxdb/influxdb/influxql   0.013s
ok      github.com/influxdb/influxdb/messaging  3.766s
?       github.com/influxdb/influxdb/opentsdb   [no test files]
ok      github.com/influxdb/influxdb/raft       8.218s
?       github.com/influxdb/influxdb/statik     [no test files]
?       github.com/influxdb/influxdb/test       [no test files]
?       github.com/influxdb/influxdb/tests/urlgen       [no test files]
?       github.com/influxdb/influxdb/udp        [no test files]
?       github.com/influxdb/influxdb/uuid       [no test files]

real    2m21.503s
user    0m20.528s
sys     0m7.652s
```

## After the change:

```
$ time go test ./... -parallel 100
ok      github.com/influxdb/influxdb    3.605s
ok      github.com/influxdb/influxdb/admin      0.012s
ok      github.com/influxdb/influxdb/client     0.011s
ok      github.com/influxdb/influxdb/cmd/influx 3.458s
ok      github.com/influxdb/influxdb/cmd/influxd        57.129s
ok      github.com/influxdb/influxdb/collectd   0.012s
ok      github.com/influxdb/influxdb/graphite   0.006s
ok      github.com/influxdb/influxdb/httpd      1.497s
ok      github.com/influxdb/influxdb/influxql   0.012s
ok      github.com/influxdb/influxdb/messaging  3.739s
?       github.com/influxdb/influxdb/opentsdb   [no test files]
ok      github.com/influxdb/influxdb/raft       7.043s
?       github.com/influxdb/influxdb/statik     [no test files]
?       github.com/influxdb/influxdb/test       [no test files]
?       github.com/influxdb/influxdb/tests/urlgen       [no test files]
?       github.com/influxdb/influxdb/udp        [no test files]
?       github.com/influxdb/influxdb/uuid       [no test files]

real    0m59.438s
user    0m20.251s
sys     0m5.510s
```

### CODE REVIEW

File: cmd/influx/main_test.go [hunk 1]
```diff
@@ -137,13 +149,27 @@ func TestQuery_Auth(t *testing.T) {
 	path := tempfile()
 	defer os.Remove(path)
 
+	// spin up a server
+	config, _ := influxd.NewTestConfig()
+
```
Comment (otoolep) [inline cmd/influx/main_test.go]: Style: any reason there are multiple blank lines here?

File: cmd/influx/main_test.go [hunk 2]
```diff
@@ -236,3 +254,19 @@ func newNode(config *influxd.Config, path string) (*influxd.Node, error) {
...
+		panic(e)
+	}
+
+	parts := strings.Split(l.Addr().String(), ":")
+	p, err := strconv.Atoi(parts[len(parts)-1])
+	if err != nil {
+		panic(err)
+	}
+	return p
+
```
Comment (otoolep) [inline cmd/influx/main_test.go]: Nit: superfluous blank line.

**Comment by otoolep** [comment]
I like the change -- especially the randomization of ports.

File: cmd/influx/main_test.go [hunk 3]
```diff
@@ -236,3 +254,19 @@ func newNode(config *influxd.Config, path string) (*influxd.Node, error) {
...
+	if e != nil {
+		panic(e)
+	}
+
+	parts := strings.Split(l.Addr().String(), ":")
+	p, err := strconv.Atoi(parts[len(parts)-1])
+	if err != nil {
+		panic(err)
+	}
+	return p
```
Comment (otoolep) [inline cmd/influx/main_test.go]: There is a tiny window here where the port is free, and could be re-used by the next caller of `newPort()`, and still be used by the first caller?

Comment (corylanou) [inline cmd/influx/main_test.go]: When watching the ports log out it always went up continuously.  Could it create a race condition?  I'm not sure.  I suspect it has more to do with how the underlying platform implements it.  I will ask on Gopher slack though to see if anyone knows if there is some type of delay at the system level before it puts that port back on the market.

Comment (otoolep) [inline cmd/influx/main_test.go]: Unfortunately we confirmed that on Ubuntu 14.04, port collision occur in this scenario.

File: cmd/influxd/run.go [hunk 4]
```diff
@@ -45,8 +45,38 @@ type Node struct {
...
+	apiListener     net.Listener      // The API TCP listener
+	GraphiteServers []graphite.Server // The Graphite Servers
+	OpenTSDBServer  *opentsdb.Server  // The OpenTSDB Server
+}
+
+func (s *Node) ClusterAddr() net.Addr {
+	return s.clusterListener.Addr()
+}
+
+func (s *Node) ClusterURL() *url.URL {
```
Comment (influxdb-denver-pair) [inline cmd/influxd/run.go]: Not super happy with this at all.  I'm open to suggestions.

Comment (corylanou) [inline cmd/influxd/run.go]: I'm not super excited about this.  I'm open to suggestions.

Comment (otoolep) [inline cmd/influxd/run.go]: How about?

```
u := url.URL{
    Scheme: "http",
    host: s.ClusterAddr().String(),
}
```

File: cmd/influxd/config.go [hunk 5]
```diff
@@ -476,21 +475,8 @@ type Graphite struct {
...
-
-	port := g.Port
-	// If no port specified, use default.
-	if port == 0 {
-		port = graphite.DefaultGraphitePort
-	}
-
-	return fmt.Sprintf("%s:%d", addr, port)
+func (g *Graphite) ConnectionString() string {
+	return fmt.Sprintf("%s:%d", g.BindAddress, g.Port)
```
Comment (otoolep) [inline cmd/influxd/config.go]: We are trying to use `net.Join` going forward, but this will work.

File: opentsdb/opentsdb.go [hunk 6]
```diff
@@ -47,8 +47,13 @@ func NewServer(w SeriesWriter, retpol string, db string) *Server {
 	return s
 }
 
+func (s *Server) Addr() net.Addr {
+	return s.listener.Addr()
+}
+
 func (s *Server) ListenAndServe(listenAddress string) {
 	var err error
+	log.Print(listenAddress)
```
Comment (otoolep) [inline opentsdb/opentsdb.go]: Is this debug logging?

**Comment by otoolep** [comment]
Some small feedback, +1. Don't forget the changelog.

**Comment by otoolep** [comment]
All done in a different PR.

**Pull request closed without merging**